esx_service/vmdk_ops.py

```diff
@@ -847,16 +848,37 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
     if not datastore:
         datastore_url = default_datastore_url
         datastore = default_datastore
+        use_default_ds = True
     else:
         datastore_url = vmdk_utils.get_datastore_url(datastore)
+        use_default_ds = False
+
+    if use_default_ds:
+        # first check whether it has privilege to default_datastore
+        error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, datastore_url, cmd, opts)
+        # "default_datastore" is set to "__VM_DS"
+        # but no privilege exist for datastore "__VM_DS"
+        # then check wheter we have a privilege to real datastore "vm_datastore", which may be
+        # added by user
+        if error_info == auth_data_const.NO_PRIVILEGE and datastore_url == auth_data_const.VM_DS_URL:
+            error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, vm_datastore_url, cmd, opts)
+        if error_info:
+            return err(error_info)
+    else:
+        # user passsed in volume with format vol@datastore
+        # check the privilege to that datastore
+        error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, datastore_url, cmd, opts)
+        # no privilege exist for the given datastore
+        # if the given datastore is the same as vm_datastore
+        # then we can check privilege against "__VM_DS"
+        if error_info == auth_data_const.NO_PRIVILEGE and datastore == vm_datastore:
+            error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, vm_datastore_url, cmd, opts)
```

Comment:
If the check at line 870 returned NO_PRIVILEGE then why would this succeed? If datastore == vm_datastore then datastore_url ==vm_datastore_url.